Optionally add faction suffix to BuildableInfo.Icon

michaeldgg2 opened the pull request:
Using different icon in `ProductionPaletteWidget` for each faction is little bit cumbersome: it's necessary to create separate unit in sequences,change `icon` sequence and use `FactionImages` field on `RenderSprites` trait.

This makes adding faction-specific icons a bit tiresome and leads to repetitive and redundant YAML definitions.

This PR adds a new feature `ProductionPaletteWidget`, which allows changing icon name of production item using function from outside. `AddFactionSuffixLogic` then leverages this and changes sequence name by adding faction name as suffix.

Changes in ra, d2k are not necessary (each faction image is actually different).

As for TS, Here's a list of actors and sequences I changed:
- Service Depot (gadept.gdi, gadept.nod)
- Light Infantry (e1.gdi, e1.nod)
- Engineer (engineer.gdi, engineer.nod)
- Tiberium Refinery (proc.gdi, proc.nod)
- Silo (gasilo.gdi, gasilo.nod)
- Service Depot (gadept.gdi, gadept.nod)

There's one building that has different sequence configuration for each faction and that is EMP Cannon. `napuls.nod` has extra `IgnoreWorldTint` field:
```
napuls.nod:
	Defaults:
		TilesetFilenames:
        ...
	emp-overlay:
		Filename: emp_fx01.shp
		TilesetFilenames:
		Length: *
		Offset: 0, 0, 25
		ZOffset: 512
		BlendMode: Additive
		IgnoreWorldTint: True
		IgnoreWorldTint: True
		DepthSprite:
	icon:
		Filename: sidebar-nod|empicon.shp
		TilesetFilenames:
		Offset: 0, 0
		DepthSprite:
```

I'm not sure what should I do here. Should I ignore `IgnoreWorldTint` field and merge `napuls.nod` and `napuls.gdi` like other sequence defintions?

PunkPun reviewed (commented):
I'd squish the 1st commit into the 2nd and 3rd into the 4th

PunkPun reviewed (dismissed):
Other than that LGTM

IceReaper commented:
LGTM :)

PunkPun commented on mods/ts/sequences/infantry.yaml:
```diff
@@ -109,18 +103,13 @@ e3:
 	Inherits: ^BasicInfantry
 	Defaults:
 		Filename: engineer.shp
 	icon:
```
```suggestion
	icon.gdi:
```

Shouldn't this be have gdi postfix then? Same with the rest of the sequences?

Though to enable that we'd have to update `CheckSequences` lint test. It's currently unaware of postfixes

PunkPun commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -484,7 +486,7 @@ public void RefreshIcons()
 				var icon = new Animation(World, rsi.GetImage(item, faction));
 				var bi = item.TraitInfo<BuildableInfo>();
-				icon.Play(bi.Icon);
+				icon.Play(TransformItemIconName(icon.Name, bi.Icon));
```
To enable linting this will likely need to be implemented deeper in the code

michaeldgg2 committed:
ts mod: Replace FactionImages usage with icon.nod sequence (AddIconFactionSuffix)

michaeldgg2 committed:
BuildableInfo provides method for constructing name of actor's production icon

michaeldgg2 committed:
Linting: faction suffix in BuildableInfo.Icon is now lintable

michaeldgg2 commented:
As per discussion with @PunkPun on Discord, I've reimplemented the icon changing feature and now the logic is inside `BuildableInfo`. This solution also makes faction-specific icons fully lintable (and thus less error-prone).

PunkPun commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -37,6 +37,9 @@ void ILintSequencesPass.Run(Action<string> emitError, Action<string> emitWarning
 		static void Run(Action<string> emitError, Action<string> emitWarning, Ruleset rules, SequenceSet sequences)
 		{
 			var factions = rules.Actors[SystemActors.World].TraitInfos<FactionInfo>().Select(f => f.InternalName).ToArray();
+			var selectableFactions = rules.Actors[SystemActors.World].TraitInfos<FactionInfo>()
```
perhaps move world actor to a variable so we only query once

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -62,6 +66,22 @@ public static string GetInitialFaction(ActorInfo ai, string defaultFaction)
 			return ai.TraitInfoOrDefault<BuildableInfo>()?.ForceFaction ?? defaultFaction;
 		}
+
+		public string GetIconSequence(Player renderPlayer, string image)
```
I think it would make a bit more sense to pass `World` and faction string rather than a `Player` reference

PunkPun replied:
Also, as this will be called every tick for every actor on production palette, it might be worth to somehow cache this. I'd assume that searching sequences and chrome metrics is not exactly cheap. I'm not sure how or if we even want to optimise though

michaeldgg2 replied:
As for `ChromeMetrics` that should be fine as retrieval using `TryGet<T>` is cached using `Dictionary`:

```
public static bool TryGet<T>(string key, out T result)
{
	if (!data.TryGetValue(key, out var s))
	{
		result = default;
		return false;
	}

	result = FieldLoader.GetValue<T>(key, s);
	return true;
}
```

Same goes to `SequenceSet` (field `images` is a `Dictionary` of `Dictionary`). Though I admit, I haven't done any profiling to be sure.

Even if these sequence names would be cached in for example `ProductionPaletteWidget` (or even in `BuildableInfo` directly), it would likely still be a `Dictionary` with key containing probably `Player.InternalName` and sequence/image name.

PunkPun replied:
I meant that as these collections aren't ordered the system needs to search all entries and there can be loads and loads of them, specifically sequences

michaeldgg2 replied:
`Dictionary<TKey, TValue>` is implementation of [Hashtable](https://en.wikipedia.org/wiki/Hashtable). This means that the search is O(1) operation, so it's definitely quick (there's no iteration through entire dictionary to find an item). Here's performance comparison between `Array` and `Dictionary`: https://gamedev.center/array-vs-dictionary-lookup-in-net/

PunkPun replied:
I suppose it's fine

PunkPun commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -98,6 +101,30 @@ static void Run(Action<string> emitError, Action<string> emitWarning, Ruleset ru
+													emitError($"Actor type `{actorInfo.Value.Name}` trait `{traitName}` field `{field.Name}` references faction specific sequence `{fullSequence}` on image `{i}`, which does not exist.");
+											}
+
+											foreach (var unkownSequence in sequencesWithPrefix)
```
```suggestion
											foreach (var unknownSequence in sequencesWithPrefix)
```

typo

PunkPun commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -98,6 +101,30 @@ static void Run(Action<string> emitError, Action<string> emitWarning, Ruleset ru
+											foreach (var unkownSequence in sequencesWithPrefix)
+											{
+												var suffix = unkownSequence[(sequence.Length + 1)..];
+												emitWarning($"Actor type `{actorInfo.Value.Name}` trait `{traitName}` field `{field.Name}` expects faction specific sequences with prefix `{sequence}` on image `{i}`, but suffix '{suffix}' does not belong to any faction.");
```
```suggestion
												emitWarning($"Actor type `{actorInfo.Value.Name}` trait `{traitName}` field `{field.Name}` has invalid faction specific sequence `{unknownSequence}` on image `{i}`, faction '{suffix}' is not a selectable faction.");
```

 I think this wording is clearer. Might also bee good to rename `suffix` to `faction`

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -35,10 +36,13 @@ public class BuildableInfo : TraitInfo<Buildable>
 		[Desc("Sequence of the actor that contains the icon.")]
 		public readonly string Icon = "icon";
 
+		[Desc("Should sequence name for " + nameof(Icon) + " have faction as suffix.")]
```
```suggestion
		[Desc("Should faction suffix be added for sequence " + nameof(Icon) + "?")]
```

PunkPun reviewed (commented):
this looks much better, but there's still a few small issues

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -58,10 +63,35 @@ public class BuildableInfo : TraitInfo<Buildable>
+		{
+			if (UseFactionSuffix)
+			{
+				if (!validFactions.Contains(faction))
```
I don't think we need a check at runtime, we have the linter for it. It's a pretty hot path. And this can both fail silently or crash depending on sequence definitions, both of the outcomes are fine here IMO

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -35,10 +37,13 @@ public class BuildableInfo : TraitInfo<Buildable>
 		public readonly string Icon = "icon";
 
+		[Desc("Should faction suffix be added for sequence " + nameof(Icon) + "?")]
+		public readonly bool UseFactionSuffix = true;
```
I'm also wondering wether this should be true by default. Base mods besides TS doesn't use this feature

PunkPun reviewed (commented):
4th commit should be merged into the 2nd, and 1st commit moved above the current 2nd

Otherwise LGTM

PunkPun commented on OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs:
```diff
@@ -134,10 +134,11 @@ public override void Draw()
+				var image = rsi.GetImage(actor, faction);
+				var icon = new Animation(world, image);
 				var bi = actor.TraitInfo<BuildableInfo>();
+				icon.Play(bi.GetIconSequence(world, GetPlayer().Faction.InternalName, image));
```
```suggestion
				icon.Play(bi.GetIconSequence(world, player.Faction.InternalName, image));
```

the variable is already cached

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -35,10 +36,13 @@ public class BuildableInfo : TraitInfo<Buildable>
 		public readonly string Icon = "icon";
 
+		[Desc("Should faction suffix be added for sequence " + nameof(Icon) + "?")]
+		public readonly bool UseFactionSuffix = false;
```
```suggestion
		public readonly bool AddIconFactionSuffix = false;
```

Perhaps something like this makes more sense? Looking at yaml the variable is unclear otherwise

PunkPun commented on OpenRA.Mods.Common/Traits/Buildable.cs:
```diff
@@ -69,7 +69,7 @@ public static string GetInitialFaction(ActorInfo ai, string defaultFaction)
 		public string GetIconSequence(World world, string faction, string image)
 		{
-			if (UseFactionSuffix)
+			if (AddIconFactionSuffix)
```
these 2 changes are in the wrong commit

abcdefg30 commented on OpenRA.Game/Traits/LintAttributes.cs:
```diff
@@ -53,14 +53,16 @@ public sealed class SequenceReferenceAttribute : Attribute
 		public readonly bool Prefix;
 		public readonly bool AllowNullImage;
 		public readonly LintDictionaryReference DictionaryReference;
+		public readonly bool FactionSuffix;
```
```suggestion
		public readonly bool HasFactionSuffix;
```

abcdefg30 commented on OpenRA.Game/Traits/LintAttributes.cs:
```diff
@@ -53,14 +53,16 @@ public sealed class SequenceReferenceAttribute : Attribute
 
 		public SequenceReferenceAttribute(string imageReference = null, bool prefix = false, bool allowNullImage = false,
-			LintDictionaryReference dictionaryReference = LintDictionaryReference.None)
+			bool factionSuffix = false, LintDictionaryReference dictionaryReference = LintDictionaryReference.None)
```
```suggestion
			bool hasFactionSuffix = false, LintDictionaryReference dictionaryReference = LintDictionaryReference.None)
```

abcdefg30 commented on mods/ts/rules/misc.yaml:
```diff
@@ -32,6 +32,8 @@ harv.colorpicker:
 	-Buildable:
 	-MapEditorData:
+	RenderSprites:
+		Image: harv
```
Why does the colorpicker actor need this now? Just because of linting? The `Buildable` trait was removed on this actor.

michaeldgg2 replied:
Yes, it does:

```OpenRA.Utility(1,1): Error: Image `harv.colorpicker` does not have any sequences defined. (Actor type `harv.colorpicker`)```

Before (the PR changes) the `RenderSprites` trait was configured on `HARV` (with `Image` name), now that's gone. The only way to remove this from `harv.colorpicker` is to put it on `HARV` (in `shared-vehicles.yaml`) - at least the only way that I know of. Do you want me to put it there?

abcdefg30 replied:
Ah no, the change makes sense to me now, thanks.

PunkPun commented:
> I'm not sure what should I do here. Should I ignore IgnoreWorldTint field and merge napuls.nod and napuls.gdi like other sequence defintions?

I think it was just overlooked, I don't see anything suggesting that IgnoreWorldTint should not have been added to it in #19122 https://github.com/OpenRA/OpenRA/commit/76a10283c40926a685ab35924258850396cc4732

michaeldgg2 commented:
Updated PR:
- merged `napuls.gdi` and `napuls.nod` images (using `icon.nod` and `icon.gdi` sequences) into `napuls`
- added `AddIconFactionSuffix: true` to `Buildable` of `NAPULS`
- `emp-overlay` sequence of `napuls` has `IgnoreWorldTint: True`
- rebased on top of bleed

penev92 commented:
I'm sorry this took almost 9 months, but [a discussion on Discord](https://discord.com/channels/153649279762694144/388282819371204608/1173917460349595699) indicates that the preferred way of achieving what the PR wants to do is in a completely different direction.
> removing icon from the actor sequences would clean up a lot of yaml
`Buildable` can just reference an image and a sequence
...
and then you have a single image with ALL. THE. ICONS.?
...
how would faction-specific build icons work with `Buildable` referencing an `build-icons` image with `e1`/`e2`/`e3` sequences?
ideally the *-specific stuff would be at the `Buildable` level
why faction specific icons only?
what if a mod wants faction specific prerequisites?
or faction specific cost?
soooo the actor has a Buildable per faction, ~~controlled by conditions?~~
...
`BuildableInfo` doesn't need to do anything or be controlled by anyone; it just needs a `RequiresFactions` string[] and then users (like `ProductionPaletteWidget`) that have all the necessary info can filter the available `BuildableInfo`s based on the player faction

So in summary, here is what I think I understood is wanted (as an even not fully related rework of `Buildable`):
1. Have a single sequence collection (image) defined with all build icons
2. Allow for multiple `Buildable` traits on an actor type and add a way to differentiate them based on faction/prerequisite/condition.
3. Have client traits/widgets filter out based on the current world/player state (faction? player condition? player prerequisite? perhaps tileset? time of day? 😅 ).

michaeldgg2 commented:
Different solution should get its own PR, so I'm closing this one.
